lib/internal/url.js

```diff
@@ -133,14 +141,44 @@ class URLContext {
     this.query = null;
     this.fragment = null;
   }
+
+  [kClone]() {
+    return {
+      data: {
+        flags: this.flags,
+        scheme: this.scheme,
+        username: this.username,
+        password: this.password,
+        host: this.host,
+        port: this.port,
+        path: this.path,
+        query: this.query,
+        fragment: this.fragment,
+      },
+      deserializeInfo: 'internal/url:URLContext'
+    };
```

Comment:
I’m wondering if it doesn’t make sense to send the URL as a string instead? Are you worried about the overhead of parsing?

Reply:
Yeah I also think that's probably better? That would mean regular URL object don't pay the extra costs of extending JSTransferable

Reply:
Ah, sorry, I was referring to the internals of the cloning method instead.

But yes, this should definitely be benchmarked – I’m not so much worried about the extra size of the superclass, that’s probably not going to be a lot compared to the already-not-so-small size of `URL` objects, but the `URL` construction might have some perf changes here.

Reply:
> Are you worried about the overhead of parsing

Little bit but I do plan to bench mark. The biggest concern is around user experience... `postMessage(url)` yields just an empty object on the receiving side, which is not super helpful. If the cost is reasonable, then providing this to allow for better dev experience is worthwhile.

Reply:
dang... yeah, benchmarks aren't ...um... great lol

```
 url/legacy-vs-whatwg-url-parse.js method='whatwg' e=1 type='auth' withBase='true'               ***    -15.16 %       ±5.02%  ±6.71%  ±8.80%
 url/legacy-vs-whatwg-url-parse.js method='whatwg' e=1 type='dot' withBase='false'               ***    -21.07 %       ±8.32% ±11.18% ±14.76%
 url/legacy-vs-whatwg-url-parse.js method='whatwg' e=1 type='dot' withBase='true'                ***    -21.09 %       ±4.53%  ±6.08%  ±8.03%
 url/legacy-vs-whatwg-url-parse.js method='whatwg' e=1 type='file' withBase='false'              ***    -18.28 %       ±8.13% ±10.86% ±14.24%
 url/legacy-vs-whatwg-url-parse.js method='whatwg' e=1 type='file' withBase='true'               ***    -15.81 %       ±4.29%  ±5.76%  ±7.61%
 url/legacy-vs-whatwg-url-parse.js method='whatwg' e=1 type='idn' withBase='false'               ***    -15.27 %       ±8.33% ±11.20% ±14.79%
 url/legacy-vs-whatwg-url-parse.js method='whatwg' e=1 type='idn' withBase='true'                ***    -17.98 %       ±3.73%  ±4.99%  ±6.54%
 url/legacy-vs-whatwg-url-parse.js method='whatwg' e=1 type='javascript' withBase='false'        ***    -22.83 %       ±8.02% ±10.76% ±14.18%
 url/legacy-vs-whatwg-url-parse.js method='whatwg' e=1 type='javascript' withBase='true'         ***    -23.51 %       ±5.07%  ±6.83%  ±9.06%
 url/legacy-vs-whatwg-url-parse.js method='whatwg' e=1 type='long' withBase='false'                      -2.09 %       ±5.65%  ±7.55%  ±9.90%
 url/legacy-vs-whatwg-url-parse.js method='whatwg' e=1 type='long' withBase='true'                **     -7.19 %       ±4.61%  ±6.18%  ±8.14%
 url/legacy-vs-whatwg-url-parse.js method='whatwg' e=1 type='percent' withBase='false'           ***    -22.24 %       ±7.26%  ±9.73% ±12.78%
 url/legacy-vs-whatwg-url-parse.js method='whatwg' e=1 type='percent' withBase='true'            ***    -21.85 %       ±2.53%  ±3.39%  ±4.47%
 url/legacy-vs-whatwg-url-parse.js method='whatwg' e=1 type='short' withBase='false'             ***    -13.37 %       ±6.94%  ±9.30% ±12.23%
 url/legacy-vs-whatwg-url-parse.js method='whatwg' e=1 type='short' withBase='true'              ***    -22.42 %       ±4.90%  ±6.54%  ±8.57%
 url/legacy-vs-whatwg-url-parse.js method='whatwg' e=1 type='wpt' withBase='false'               ***    -15.30 %       ±5.60%  ±7.48%  ±9.81%
 url/legacy-vs-whatwg-url-parse.js method='whatwg' e=1 type='wpt' withBase='true'                ***    -18.42 %       ±6.74%  ±9.05% ±11.93%
 url/legacy-vs-whatwg-url-parse.js method='whatwg' e=1 type='ws' withBase='false'                ***    -22.09 %       ±7.86% ±10.54% ±13.90%
 url/legacy-vs-whatwg-url-parse.js method='whatwg' e=1 type='ws' withBase='true'                 ***    -21.10 %       ±3.79%  ±5.07%  ±6.64%
```

Reply:
@addaleax... Just thinking through an alternative approach to handling object cloning so that we don't have to extend JSTransferable every time and incur this kind of perf hit... what if we defined an internal symbol that classes could implement that would acquire a Transferable object delegate?

e.g. something like...

```js
const {
  kTransferableDelegate,
  JSTransferable
} = require('internal/...')

class URLTransferDelegate extends JSTransferable {
  // ...
}

class URL {
  // ...

  [kTransferableDelegate]() {
    return new URLTransferDelegate(this);
  }
}
```

Here we can limit the cost of extracting the delegate to inside thhe PostMessage implementation without incurring an additional cost on the source objects themselves.

Reply:
I think this doesn't have to be polymoryhic though - can't post message see it got a URL instance and handle cloning which would mean 0 cost for URLs?

Reply:
The way V8 currently works is that it can only do special handling for host objects, i.e. ones with internal fields. A symbol-based approach would be nice, but is just not possible currently, otherwise I would have implemented that instead. 

> can't post message see it got a URL instance

That would imply walking the object tree inside `postMessage`. I think that’s a non-starter.

Reply:
It's not ideal, but we could expose a function or class to users who want to transfer URLs?
Something like `makeTransferableURL(new URL('...'))` or `new TransferableURL(new URL(''))` so the cost is not payed by everyone?

Related to this, have we already considered exposing somehow a way to make any custom object transferable by the users?

Reply:
> It's not ideal, but we could expose a function or class to users who want to transfer URLs?

That’s definitely possible, yes.

> Related to this, have we already considered exposing somehow a way to make any custom object transferable by the users?

That would also need support from V8 first, I think. What we could do is add functionality like this to N-API objects at some point, which I think would be a decent feature to offer.

Reply:
> That would imply walking the object tree inside postMessage. I think that’s a non-starter.

What if it's opt-in? A `transferable` method that goes and replaces all the variants of objects with their transferable variants (that are basically the same but extends JSTransferable)

Reply:
@benjamingr I think that still comes with a lot of complexity … I think if somebody wants to put in the work, that’s okay, but I still think it’s a lot of complexity for adding a nonstandard feature. Maybe it’s worth asking the V8 team if they want to support some generic hook for this, e.g. symbol-based?

Reply:
Yes, that would be ideal.

Reply:
> What we could do is add functionality like this to N-API objects at some point

I've started investigating this. It's non-trivial but workable. The key challenge is that the current napi_define_class and ObjectWrap APIs do not treat the objects as host objects. Not insurmountable tho and doesn't help us in this case. 

I'll try the symbol based approach I suggested in https://github.com/nodejs/node/pull/36434#discussion_r537957620 and see where that gets us in terms of performance. In general, however, V8 support would be ideal.